### ISSUE TITLE
Add ability to capture unbarcoded versions of RDTs: Alltest, Wondfo, GreenSpring

### ISSUE THREAD

**Issue by shivenbhatt1** [onaio/rdt-standard#332]
This can be added to the barcoded/manual entry screen for RDT or added as a dropdown to the manual entry field for RDTs. All tests, except Greenspring have already been integrated by UW.

**Comment by pld** [comment]
Need scoped

**Comment by shivenbhatt1** [comment]
@pld @vincent-karuri Right now there are two options 1) barcode RDT which has a pre-identified RDT and associated template image and 2) manual RDT capture which does not have a pre-identified template image for RDTs used without barcoded information

We'd like to add the preexisting known RDTs that already have existing template images in the UW library to the "Enter RDT manually" field.

- [ ] Add RDTs that exist in UW code to the top of the Record RDT Information form (i.e. create a dropdown field for choose RDT type, have Alltest and Wondfo available - if we integrate more, they should be added, and add other)

If the user selects other, they are prompted to fill out the manufacturer name and RDT type, batch ID and expiration date

If the user selects an existing manufacturer from the dropdown, RDT type is not needed, but the user still needs to add RDT batch ID and RDT expiration date on this form.

Just a side note, it would be nice to have template images of the RDT here, as discussed previously, display so that the user can confirm that the RDT they chose is the same RDT.

**Comment by shivenbhatt1** [comment]
need a config file for RDT type, and the test choreography for it

**Comment by pld** [comment]
Cool, sorry I dropped off @vincent-karuri definitely need your inputs on this, putting my initial thoughts below.

What we want to do here is use a configuration file to generate the list of RDTs that users can select. This configuration file will be used on the input selection screen, giving users the option of barcode scan, manual, or all the items in this list. Later we could move the list to be backed by the db and integrating with existing form configuration work (we need to understand how this integrates with existing form config work better).

Do we use a particular schema standard (like [JSON schema](https://json-schema.org/specification.html#specification-documents) or something) in the form configuration library? Here's example content (not schema) for the file:

```json
[
    {
        "name": "Wondfo",
        "conditions": [
            {
                "name": "Covid-19",
                "type": "IgG/IgM"
            },
        ],
        "testTime": "10 min",
        "gtin": "XXXX",
        "rdtReaderTemplate": "[B64 IMAGE DATA]",
        "instructionImage": "[B64 IMAGE DATA]",
        "instructionText": "Enter buffer into buffer well"
    },
    {
        "name": "Green Spring",
        "conditions": [
            {
                "name": "Covid-19",
                "type": "IgG/IgM"
            },
            {
                "name": "Covid-19",
                "type": "Antigen"
            },
        ],
        "testTime": "15 min",
        "gtin": "XXXX",
        "rdtReaderTemplate": "[B64 IMAGE DATA]",
        "instructionImage": "[B64 IMAGE DATA]",
        "instructionText": "Enter buffer into buffer well"
    }
]
```

@shivenbhatt1 we definitely need to think through this schema more and confirm it with Anu. Is there other info you can think of to add to this? Reassigning the issue to you for comments and coordination of next conversation about this.

**Comment by vincent-karuri** [comment]
As per an earlier discussion with @ekigamba any valid `JSON` is accepted as an asset file by the form configuration library.

The `instructionText` and `instructionImage` may need to be an array of `JSONObject`s, considering we may want multiple instruction screen steps. We could also just have a single field that references a sub-form containing the flow of instructions. 

Not sure we need to embed the binary image in the config? Could we separate it into it's own file and reference the file by introducing a `JSON` field instead?

**Comment by pld** [comment]
> As per an earlier discussion with @ekigamba any valid `JSON` is accepted as an asset file by the form configuration library.
> 
> The `instructionText` and `instructionImage` may need to be an array of `JSONObject`s, considering we may want multiple instruction screen steps. We could also just have a single field that references a sub-form containing the flow of instructions.
> 
> Not sure we need to embed the binary image in the config? Could we separate it into it's own file and reference the file by introducing a `JSON` field instead?

yea good point, we could change the instruction keys to a list of steps:

```json
{
    "instructionsSteps": [
        {
            "instructionText": "Step 1: Fill sample",
            "instructionImage": "[B64 data]",
        },
        {
            "instructionText": "Step 2: Fill buffer",
            "instructionImage": "[B64 data]",
        }
    ]
}
```

I'm not against a reference to files instead of embed, this seemed simpler, to avoid storing, syncing, loading external assets

**Comment by pld** [comment]
I’ve been looking through the FHIR docs, starting here http://www.hl7.org/FHIR/diagnostics-module.html, to try and find the closest module, so far DiagnosticReport, http://www.hl7.org/FHIR/diagnosticreport.html

**Comment by craigappl** [comment]
Check to see if there are any limitations to the DiagnosticReport. The Hamilton Health Sciences team in this implementation guide chose to send a bundle with a DocumentReference: https://simplifier.net/guide/LungCancerDiagnosisComputableCareGuidline/Resources I could send them an email to see why they chose that instead of a DiagnosticReport.

**Comment by pld** [comment]
this is much closer, http://www.hl7.org/FHIR/device.html

**Comment by pld** [comment]
**tl;dr define each RDT we want the open reader to accommodate by writing a FHIR `DeviceDefinition`**


oh sweet this looks like EXACTLY what we want, http://www.hl7.org/FHIR/devicedefinition.html, this is the schema

```javascript
{
  "resourceType" : "DeviceDefinition",
  // from Resource: id, meta, implicitRules, and language
  // from DomainResource: text, contained, extension, and modifierExtension
  "identifier" : [{ Identifier }], // Instance identifier
  "udiDeviceIdentifier" : [{ // Unique Device Identifier (UDI) Barcode string
    "deviceIdentifier" : "<string>", // R!  The identifier that is to be associated with every Device that references this DeviceDefintiion for the issuer and jurisdication porvided in the DeviceDefinition.udiDeviceIdentifier
    "issuer" : "<uri>", // R!  The organization that assigns the identifier algorithm
    "jurisdiction" : "<uri>" // R!  The jurisdiction to which the deviceIdentifier applies
  }],
  // manufacturer[x]: Name of device manufacturer. One of these 2:
  "manufacturerString" : "<string>",
  "manufacturerReference" : { Reference(Organization) },
  "deviceName" : [{ // A name given to the device to identify it
    "name" : "<string>", // R!  The name of the device
    "type" : "<code>" // R!  udi-label-name | user-friendly-name | patient-reported-name | manufacturer-name | model-name | other
  }],
  "modelNumber" : "<string>", // The model number for the device
  "type" : { CodeableConcept }, // What kind of device or device system this is
  "specialization" : [{ // The capabilities supported on a  device, the standards to which the device conforms for a particular purpose, and used for the communication
    "systemType" : "<string>", // R!  The standard that is used to operate and communicate
    "version" : "<string>" // The version of the standard that is used to operate and communicate
  }],
  "version" : ["<string>"], // Available versions
  "safety" : [{ CodeableConcept }], // Safety characteristics of the device
  "shelfLifeStorage" : [{ ProductShelfLife }], // Shelf Life and storage information
  "physicalCharacteristics" : { ProdCharacteristic }, // Dimensions, color etc.
  "languageCode" : [{ CodeableConcept }], // Language code for the human-readable text strings produced by the device (all supported)
  "capability" : [{ // Device capabilities
    "type" : { CodeableConcept }, // R!  Type of capability
    "description" : [{ CodeableConcept }] // Description of capability
  }],
  "property" : [{ // The actual configuration settings of a device as it actually operates, e.g., regulation status, time properties
    "type" : { CodeableConcept }, // R!  Code that specifies the property DeviceDefinitionPropetyCode (Extensible)
    "valueQuantity" : [{ Quantity }], // Property value as a quantity
    "valueCode" : [{ CodeableConcept }] // Property value as a code, e.g., NTP4 (synced to NTP)
  }],
  "owner" : { Reference(Organization) }, // Organization responsible for device
  "contact" : [{ ContactPoint }], // Details for human/organization for support
  "url" : "<uri>", // Network address to contact device
  "onlineInformation" : "<uri>", // Access to on-line information
  "note" : [{ Annotation }], // Device notes and comments
  "quantity" : { Quantity }, // The quantity of the device present in the packaging (e.g. the number of devices present in a pack, or the number of devices in the same package of the medicinal product)
  "parentDevice" : { Reference(DeviceDefinition) }, // The parent device it can be part of
  "material" : [{ // A substance used to create the material(s) of which the device is made
    "substance" : { CodeableConcept }, // R!  The substance
    "alternate" : <boolean>, // Indicates an alternative material of the device
    "allergenicIndicator" : <boolean> // Whether the substance is a known or suspected allergen
  }]
}
```

---

here's an example, for our context filled out w/some stuff I could find:

```javascript
{
  "resourceType" : "DeviceDefinition",
  // from Resource: id, meta, implicitRules, and language
  // from DomainResource: text, contained, extension, and modifierExtension
  "identifier" : "[UUID we generate]", // Instance identifier
  "udiDeviceIdentifier" : [{ // Unique Device Identifier (UDI) Barcode string
    "deviceIdentifier" : "[GTIN]", // R!  The identifier that is to be associated with every Device that references this DeviceDefintiion for the issuer and jurisdication porvided in the DeviceDefinition.udiDeviceIdentifier
  }],
  // manufacturerString: "Wondfo",
  "deviceName" : [{ // A name given to the device to identify it
    "name" : "Wondfo Covid-19 IgG/IgM", // R!  The name of the device
    "type" : "user-friendly-name" // R!  udi-label-name | user-friendly-name | patient-reported-name | manufacturer-name | model-name | other
  }],
  "modelNumber" : "[TBD need this]", // The model number for the device
  "type" : [87717000](http://browser.ihtsdotools.org/?perspective=full&conceptId1=87717000), // What kind of device or device system this is
  "version" : ["0.1"], // Available versions
  "physicalCharacteristics" : {
    "image": "[B64 data, or attachment which this supports]"
  }, // Dimensions, color etc.
  "capability" : [{ // Device capabilities
    {
      "type": { "text": "Step 1: Fill sample },
      "description": { "text": "[B64 data]" },
    },
    {
      "type": { "text": "Step 1: Fill sample },
      "description": { "text": "[B64 data]" },
    }
  ],
  "property" : [{ // The actual configuration settings of a device as it actually operates, e.g., regulation status, time properties
      "type" : { "text": "TimeForResult" }, // R!  Code that specifies the property DeviceDefinitionPropetyCode (Extensible)
      "valueQuantity" : [{
        "value": "10",
        "unit": "minutes"
       }], // Property value as a quantity
    },
    {
      "type" : { "text": "IgG" }, // R!  Type of capability
    },
    {
      "type" : { "text": "IgM" }, // R!  Type of capability
    },
    {
      "type" : { "text": "rdtReaderTemplate" }, // R!  Type of capability
      "valueCode: { "text": "[B64 data]" }
    }

  ],
}
```

@shivenbhatt1 this is the content of that definition, does this look good to you? we'll want to create a record matching this schema for each device we want to support:

I did NOT define a `parentDevice` it makes sense to have `RDT-OG` as a parent device that all the RDT-OG conforming devices inherit from.

It's kinda janky how I hacked the instructions and the template images into this, but at least it stays in the format, making our own resource *might* make sense, but before that we should think if there's a way to improve on the way I put this together within the existing structure.

reassigning this to @shivenbhatt1 and @vincent-karuri, I think we're set to move forward

**Comment by pld** [comment]
Relatedly, but not needed to continue on this issue, I worked up a the FHIR Resources the RDT-OG will be organized around here, https://docs.google.com/drawings/d/1gClpMT6fvFOpE6_arznV-uBsKfu6tKqmIeULiBjoWEQ/edit, will move this into the standard doc

**Comment by shivenbhatt1** [comment]
> **tl;dr define each RDT we want the open reader to accommodate by writing a FHIR `DeviceDefinition`**
> 
> oh sweet this looks like EXACTLY what we want, http://www.hl7.org/FHIR/devicedefinition.html, this is the schema
> 
> ```js
> {
>   "resourceType" : "DeviceDefinition",
>   // from Resource: id, meta, implicitRules, and language
>   // from DomainResource: text, contained, extension, and modifierExtension
>   "identifier" : [{ Identifier }], // Instance identifier
>   "udiDeviceIdentifier" : [{ // Unique Device Identifier (UDI) Barcode string
>     "deviceIdentifier" : "<string>", // R!  The identifier that is to be associated with every Device that references this DeviceDefintiion for the issuer and jurisdication porvided in the DeviceDefinition.udiDeviceIdentifier
>     "issuer" : "<uri>", // R!  The organization that assigns the identifier algorithm
>     "jurisdiction" : "<uri>" // R!  The jurisdiction to which the deviceIdentifier applies
>   }],
>   // manufacturer[x]: Name of device manufacturer. One of these 2:
>   "manufacturerString" : "<string>",
>   "manufacturerReference" : { Reference(Organization) },
>   "deviceName" : [{ // A name given to the device to identify it
>     "name" : "<string>", // R!  The name of the device
>     "type" : "<code>" // R!  udi-label-name | user-friendly-name | patient-reported-name | manufacturer-name | model-name | other
>   }],
>   "modelNumber" : "<string>", // The model number for the device
>   "type" : { CodeableConcept }, // What kind of device or device system this is
>   "specialization" : [{ // The capabilities supported on a  device, the standards to which the device conforms for a particular purpose, and used for the communication
>     "systemType" : "<string>", // R!  The standard that is used to operate and communicate
>     "version" : "<string>" // The version of the standard that is used to operate and communicate
>   }],
>   "version" : ["<string>"], // Available versions
>   "safety" : [{ CodeableConcept }], // Safety characteristics of the device
>   "shelfLifeStorage" : [{ ProductShelfLife }], // Shelf Life and storage information
>   "physicalCharacteristics" : { ProdCharacteristic }, // Dimensions, color etc.
>   "languageCode" : [{ CodeableConcept }], // Language code for the human-readable text strings produced by the device (all supported)
>   "capability" : [{ // Device capabilities
>     "type" : { CodeableConcept }, // R!  Type of capability
>     "description" : [{ CodeableConcept }] // Description of capability
>   }],
>   "property" : [{ // The actual configuration settings of a device as it actually operates, e.g., regulation status, time properties
>     "type" : { CodeableConcept }, // R!  Code that specifies the property DeviceDefinitionPropetyCode (Extensible)
>     "valueQuantity" : [{ Quantity }], // Property value as a quantity
>     "valueCode" : [{ CodeableConcept }] // Property value as a code, e.g., NTP4 (synced to NTP)
>   }],
>   "owner" : { Reference(Organization) }, // Organization responsible for device
>   "contact" : [{ ContactPoint }], // Details for human/organization for support
>   "url" : "<uri>", // Network address to contact device
>   "onlineInformation" : "<uri>", // Access to on-line information
>   "note" : [{ Annotation }], // Device notes and comments
>   "quantity" : { Quantity }, // The quantity of the device present in the packaging (e.g. the number of devices present in a pack, or the number of devices in the same package of the medicinal product)
>   "parentDevice" : { Reference(DeviceDefinition) }, // The parent device it can be part of
>   "material" : [{ // A substance used to create the material(s) of which the device is made
>     "substance" : { CodeableConcept }, // R!  The substance
>     "alternate" : <boolean>, // Indicates an alternative material of the device
>     "allergenicIndicator" : <boolean> // Whether the substance is a known or suspected allergen
>   }]
> }
> ```
> 
> here's an example, for our context filled out w/some stuff I could find:
> 
> ```js
> {
>   "resourceType" : "DeviceDefinition",
>   // from Resource: id, meta, implicitRules, and language
>   // from DomainResource: text, contained, extension, and modifierExtension
>   "identifier" : "[UUID we generate]", // Instance identifier
>   "udiDeviceIdentifier" : [{ // Unique Device Identifier (UDI) Barcode string
>     "deviceIdentifier" : "[GTIN]", // R!  The identifier that is to be associated with every Device that references this DeviceDefintiion for the issuer and jurisdication porvided in the DeviceDefinition.udiDeviceIdentifier
>   }],
>   // manufacturerString: "Wondfo",
>   "deviceName" : [{ // A name given to the device to identify it
>     "name" : "Wondfo Covid-19 IgG/IgM", // R!  The name of the device
>     "type" : "user-friendly-name" // R!  udi-label-name | user-friendly-name | patient-reported-name | manufacturer-name | model-name | other
>   }],
>   "modelNumber" : "[TBD need this]", // The model number for the device
>   "type" : [87717000](http://browser.ihtsdotools.org/?perspective=full&conceptId1=87717000), // What kind of device or device system this is
>   "version" : ["0.1"], // Available versions
>   "physicalCharacteristics" : {
>     "image": "[B64 data, or attachment which this supports]"
>   }, // Dimensions, color etc.
>   "capability" : [{ // Device capabilities
>     {
>       "type": { "text": "Step 1: Fill sample },
>       "description": { "text": "[B64 data]" },
>     },
>     {
>       "type": { "text": "Step 1: Fill sample },
>       "description": { "text": "[B64 data]" },
>     }
>   ],
>   "property" : [{ // The actual configuration settings of a device as it actually operates, e.g., regulation status, time properties
>       "type" : { "text": "TimeForResult" }, // R!  Code that specifies the property DeviceDefinitionPropetyCode (Extensible)
>       "valueQuantity" : [{
>         "value": "10",
>         "unit": "minutes"
>        }], // Property value as a quantity
>     },
>     {
>       "type" : { "text": "IgG" }, // R!  Type of capability
>     },
>     {
>       "type" : { "text": "IgM" }, // R!  Type of capability
>     },
>     {
>       "type" : { "text": "rdtReaderTemplate" }, // R!  Type of capability
>       "valueCode: { "text": "[B64 data]" }
>     }
> 
>   ],
> }
> ```
> 
> @shivenbhatt1 this is the content of that definition, does this look good to you? we'll want to create a record matching this schema for each device we want to support:
> 
> I did NOT define a `parentDevice` it makes sense to have `RDT-OG` as a parent device that all the RDT-OG conforming devices inherit from.
> 
> It's kinda janky how I hacked the instructions and the template images into this, but at least it stays in the format, making our own resource _might_ make sense, but before that we should think if there's a way to improve on the way I put this together within the existing structure.
> 
> reassigning this to @shivenbhatt1 and @vincent-karuri, I think we're set to move forward

@pld This looks good, I see the add sample, add buffer, time to capture. I think we are just missing the boundary of time, i.e. ensuring that the user captures between the time of 10 to 30 mins when the test result is reliable, and anything after 30 mins is outside the protocol.

**Comment by pld** [comment]
cool, above I put time in the `"property"` key, below I changed that to be minimum and maximum time:

```
{
  "resourceType" : "DeviceDefinition",
  // from Resource: id, meta, implicitRules, and language
  // from DomainResource: text, contained, extension, and modifierExtension
  "identifier" : "[UUID we generate]", // Instance identifier
  "udiDeviceIdentifier" : [{ // Unique Device Identifier (UDI) Barcode string
    "deviceIdentifier" : "[GTIN]", // R!  The identifier that is to be associated with every Device that references this DeviceDefintiion for the issuer and jurisdication porvided in the DeviceDefinition.udiDeviceIdentifier
  }],
  // manufacturerString: "Wondfo",
  "deviceName" : [{ // A name given to the device to identify it
    "name" : "Wondfo Covid-19 IgG/IgM", // R!  The name of the device
    "type" : "user-friendly-name" // R!  udi-label-name | user-friendly-name | patient-reported-name | manufacturer-name | model-name | other
  }],
  "modelNumber" : "[TBD need this]", // The model number for the device
  "type" : [87717000](http://browser.ihtsdotools.org/?perspective=full&conceptId1=87717000), // What kind of device or device system this is
  "version" : ["0.1"], // Available versions
  "physicalCharacteristics" : {
    "image": "[B64 data, or attachment which this supports]"
  }, // Dimensions, color etc.
  "capability" : [{ // Device capabilities
    {
      "type": { "text": "Step 1: Fill sample },
      "description": { "text": "[B64 data]" },
    },
    {
      "type": { "text": "Step 1: Fill sample },
      "description": { "text": "[B64 data]" },
    }
  ],
  "property" : [{ // The actual configuration settings of a device as it actually operates, e.g., regulation status, time properties
      "type" : { "text": "minimumTimeForResult" }
      "valueQuantity" : [{
        "value": "10",
        "unit": "minutes"
       }], // Property value as a quantity
    },
    {
      "type" : { "text": "maximumTimeForResult" },
      "valueQuantity" : [{
        "value": "30",
        "unit": "minutes"
       }], // Property value as a quantity
    },
    {
      "type" : { "text": "IgG" }, // R!  Type of capability
    },
    {
      "type" : { "text": "IgM" }, // R!  Type of capability
    },
    {
      "type" : { "text": "rdtReaderTemplate" }, // R!  Type of capability
      "valueCode: { "text": "[B64 data]" }
    }

  ],
}
```

how does that look?

**Comment by shivenbhatt1** [comment]
Looks good to me. Just wondering, how do we define a test with a single line that has IgG and IgM and one that has 2 separate lines?

**Comment by pld** [comment]
I was thinking this would be for a test w/2 separate lines, for a test w/ a single line we'd replace that part of properties with

```javascript
{
  "type" : { "text": "IgG/IgM" }, // R!  Type of capability
}
```

We have leeway in how we want to do this and we can adjust after we decide on the first version but that will mean changes to the code that interprets it.

I'm expecting we will have to adjust it somewhat as we learn, we haven't tried to use this FHIR resource before.

**Comment by pld** [comment]
@shiven do we have a spreadsheet with this info for the tests we want to support?

If not can you create that and send me a link, I can check to make sure it has all the columns we need, then you can fill in the info and when that's done I'll create a calculation so it generates the FHIR resources we need for this issue and export that as a JSON file.

Finally, after have the JSON, I'll create a new implementation issue to hand over to the engineering team.
